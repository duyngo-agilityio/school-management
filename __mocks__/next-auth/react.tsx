import { createContext } from "react";

export const useSession = () => ({
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
});

export type Session = {
  data: {
    user: {
      token: string;
      email: string;
      id: string;
      name: string;
    };
    expires: string;
  };
  status: "loading" | "authenticated" | "unauthenticated";
  update: (data?: Session) => Promise<Session | null>;
};

const INITIAL_SESSION: Session = {
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
};

const SessionContext = createContext(INITIAL_SESSION);

export const SessionProvider = ({ children }: React.PropsWithChildren) => {
  <SessionContext.Provider value={INITIAL_SESSION}>
    {children}
  </SessionContext.Provider>;
};
