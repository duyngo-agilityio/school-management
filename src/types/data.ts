export interface IStudent {
  id: string;
  fullName: string;
  gender: number;
  age: number;
  className: string;
  password?: string;
  email?: string;
  description?: string;
  phoneNumber?: string;
  avatar?: string;
}

export interface ITeacher {
  id: string;
  fullName: string;
  gender: number;
  age: number;
  className: string;
  subject?: string;
  password?: string;
  email?: string;
  description?: string;
  phoneNumber?: string;
  avatar?: string;
}

export type TDataType = IStudent | ITeacher;

export type APIResponse<T> = {
  isSuccess: boolean;
  data: T | null;
};

export type TSearchParams = {
  q: string;
};