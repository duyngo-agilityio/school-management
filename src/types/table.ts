export interface IFiledNameColumns {
  field: string;
  headerName: string;
}
