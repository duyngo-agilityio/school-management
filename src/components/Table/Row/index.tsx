'use client';

import { memo, MouseEvent, ReactNode } from 'react';

// Components
import { Tr, TableRowProps as ChakraTableRowProps } from '@chakra-ui/react';
import isEqual from 'react-fast-compare';

interface TableRowProps extends ChakraTableRowProps {
  children: ReactNode;
  onClickRow?: (event: MouseEvent<HTMLTableRowElement>) => void;
}

const TableRow = ({ children, onClickRow, ...rest }: TableRowProps) => {
  return (
    <Tr onClick={onClickRow} {...rest}>
      {children}
    </Tr>
  );
};

export default memo(TableRow, isEqual);
