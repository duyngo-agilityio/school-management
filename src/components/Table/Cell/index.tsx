// Libs
import { memo } from 'react';
import isEqual from 'react-fast-compare';

// Components
import { Td } from '@chakra-ui/react';

// Types
import { IFiledNameColumns, TDataType } from '@/types';

// Lodash
import get from 'lodash.get';

interface TableCellProps {
  item: TDataType;
  column: IFiledNameColumns;
}

const TableCell = ({ item, column }: TableCellProps) => {
  const value = get(item, column.field);
  const renderCellValue = column.render ? column.render(column, item) : value;

  return (
    <Td borderBottom="none" fontSize="12px">
      {renderCellValue}
    </Td>
  );
};

export default memo(TableCell, isEqual);
