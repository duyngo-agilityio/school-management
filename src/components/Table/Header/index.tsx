// Components
import { Th, Thead, Tr } from '@chakra-ui/react';

// Types
import { IFiledNameColumns } from '@/types';

interface TableHeaderProps {
  columns: IFiledNameColumns[];
}

const TableHeader = ({ columns }: TableHeaderProps) => {
  const renderColumns = columns.map(({ field, headerName }) => (
    <Th key={field} borderBottom="none">{headerName}</Th>
  ));

  return (
    <Thead>
      <Tr>{renderColumns}</Tr>
    </Thead>
  );
};

export default TableHeader;
