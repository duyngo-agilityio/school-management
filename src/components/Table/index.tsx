'use client';

import { useState } from 'react';

// Components
import { TableContainer, Tbody, Table as ChakraTable } from '@chakra-ui/react';
import TableHeader from './Header';
import TableRow from './Row';
import TableCell from './Cell';

// Types
import { IFiledNameColumns, TDataType } from '@/types';

interface TableProps {
  columns: IFiledNameColumns[];
  data: TDataType[];
  onClickRow: (id: string) => void;
}

const Table = ({ columns, data, onClickRow }: TableProps) => {
  const [selectedId, setSelectedId] = useState<string>('');

  const handleClickRow = (id: string) => () => {
    setSelectedId(id);
    onClickRow(id);
  };

  return (
    <TableContainer>
      <ChakraTable>
        <TableHeader columns={columns} />
        <Tbody>
          {data.map((item, index) => {
            const activeRow = selectedId === item.id;
            const bgColor = activeRow
              ? 'backgroundBlueHaveLock'
              : index % 2 === 0
                ? 'backgroundSolitude'
                : '';

            const textColor = activeRow ? 'textSecondary' : 'textPrimary';

            return (
              <TableRow
                key={item.id}
                onClickRow={handleClickRow(item.id)}
                bg={bgColor}
                color={textColor}
                cursor="pointer"
                borderBottom="none"
              >
                {columns.map((column) => (
                  <TableCell key={column.field} column={column} item={item} />
                ))}
              </TableRow>
            );
          })}
        </Tbody>
      </ChakraTable>
    </TableContainer>
  );
};

export default Table;
