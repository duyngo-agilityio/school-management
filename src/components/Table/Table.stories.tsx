import type { Meta, StoryObj } from '@storybook/react';

// Types
import { TDataType } from '@/types';

// Mocks
import { COLUMNS_STUDENT, MOCK_STUDENTS } from '@/mocks';

// Components
import Table from '.';

const meta = {
  title: 'components/Table',
  component: Table,
  tags: ['autodocs'],
  argTypes: {
    columns:{ description: 'Columns are fundamental elements that play a pivotal role in rendering column values in the required format within your application'},
    data: { description: 'Array included information show in table' }
  }
} as Meta<typeof Table>;

export default meta;

type Story = StoryObj<typeof meta>;

const data: TDataType[] = MOCK_STUDENTS

export const Default: Story = {
  args: {
    columns: COLUMNS_STUDENT,
    data: data
  }
};