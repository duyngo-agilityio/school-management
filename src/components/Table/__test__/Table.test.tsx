import { render } from '@testing-library/react';
import { IFiledNameColumns, TDataType } from '@/types';

// Components
import Table from '..';

jest.mock('next/navigation', () => ({
  useParams: jest.fn(),
}));

describe('Table Component', () => {
  const columns: IFiledNameColumns[] = [
    { field: 'id', headerName: 'ID' },
    { field: 'fullName', headerName: 'Name' },
  ];

  const data: TDataType[] = [
    { id: '1', fullName: 'Item 1', age: 32, className: 'SS 1', gender: 1 },
    { id: '2', fullName: 'Item 2', age: 12, className: 'SS 2', gender: 1 },
    { id: '3', fullName: 'Item 3', age: 14, className: 'SS 3', gender: 1 },
  ];

  it('should match the snapshot', () => {
    const mockUseParams = jest.requireMock('next/navigation').useParams;
    mockUseParams.mockReturnValue({ id: '1' });

    const { container } = render(<Table columns={columns} data={data} />);
    expect(container).toMatchSnapshot();
  });

  it('should set bgColor to "backgroundBlueHaveLock" for the active row', () => {
    const mockUseParams = jest.requireMock('next/navigation').useParams;
    mockUseParams.mockReturnValue({ id: '1' });

    const { getByText } = render(<Table columns={columns} data={data} />);
    const activeRow = getByText('Item 1').closest('tr');

    expect(activeRow).toHaveStyle('background-color: backgroundBlueHaveLock');
  });

  it('should set bgColor to "backgroundSolitude" for an even index when not active', () => {
    const mockUseParams = jest.requireMock('next/navigation').useParams;
    mockUseParams.mockReturnValue({ id: '4' });

    const { getByText } = render(<Table columns={columns} data={data} />);
    const evenIndexRow = getByText('Item 2').closest('tr');

    expect(evenIndexRow).toHaveStyle('background-color: backgroundSolitude');
  });

  it('should set bgColor to an empty string for an odd index when not active', () => {
    const mockUseParams = jest.requireMock('next/navigation').useParams;
    mockUseParams.mockReturnValue({ id: '4' });

    const { getByText } = render(<Table columns={columns} data={data} />);
    const oddIndexRow = getByText('Item 3').closest('tr');

    expect(oddIndexRow).toHaveStyle('background-color: ');
  });
});
