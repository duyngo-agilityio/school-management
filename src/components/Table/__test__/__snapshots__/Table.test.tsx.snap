// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should match the snapshot 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-0"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-ujzx5o"
          >
            ID
          </th>
          <th
            class="css-ujzx5o"
          >
            Name
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-1t3b6fg"
        >
          <td
            class="css-4e3ndm"
          >
            1
          </td>
          <td
            class="css-4e3ndm"
          >
            Item 1
          </td>
        </tr>
        <tr
          class="css-bgbn71"
        >
          <td
            class="css-4e3ndm"
          >
            2
          </td>
          <td
            class="css-4e3ndm"
          >
            Item 2
          </td>
        </tr>
        <tr
          class="css-pb1fbr"
        >
          <td
            class="css-4e3ndm"
          >
            3
          </td>
          <td
            class="css-4e3ndm"
          >
            Item 3
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
